core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -536,11 +536,12 @@ private TaskInfo doUpdateTask(
                     Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
                             .map(CatalogProperties::catalogHandle)
                             .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
+                    if (sqlTask.setCatalogs(catalogHandles) || !sqlTask.catalogsLoaded()) {
```

Comment:
dio do we have two use double-check pattern here?
we get `!sqlTask.catalogsLoaded()` and then we take a lock and re-check

the whole complexity could perhaps be hidden inside `ensureCatalogsLoaded()` method.
if it cannot, the code should be commented, documenting why the extra complexity is needed

Reply:
@jklamer let's talk about simplifying this code in some follow up work.  I'm thinking we can hoist the read lock around the larger statement, and then move the entire update inside of Task by giving Task access to `connectorServicesProvider`... maybe.  It is just a thought.

Reply:
yeah, I think can rework this mechanism. The key complexity is that the catalogs being set in the SqlTask has to be ordered with the prune catalog task in this class. And then internal to the Catalog manager the ensureCatalogLoaded has to be ordered with the pruning as well, but for different reasons.
If the catalog manager was able to add the catalogs to the task we might be able to flatten it all internal to the manager

Reply:
what does it mean "has to be ordered with the prune catalog task in this class"?

Reply:
I don't think "ordered" is the right term here.  To prune catalogs, we must temporarily pause adding catalogs while we compute the which catalogs to prune.